File: example/java/jimmer-sql/repository/src/main/java/org/babyfish/jimmer/sql/example/repository/AuthorRepository.java
package org.babyfish.jimmer.sql.example.repository;

import org.babyfish.jimmer.spring.repository.JRepository;
import org.babyfish.jimmer.spring.repository.SpringOrders;
import org.babyfish.jimmer.sql.ast.query.specification.JSpecification;
import org.babyfish.jimmer.sql.example.model.Author;
import org.babyfish.jimmer.sql.example.model.AuthorTable;
import org.babyfish.jimmer.sql.example.model.Tables;
import org.babyfish.jimmer.sql.fetcher.Fetcher;
import org.springframework.data.domain.Sort;

import java.util.List;

public interface AuthorRepository extends JRepository<Author, Long>, Tables {

    AuthorTable table = AUTHOR_TABLE;

    default List<Author> find(
            JSpecification<?, AuthorTable> specification,
            Sort sort,
            Fetcher<Author> fetcher
    ) {
        return sql().createQuery(table)
                .where(specification)
                .orderBy(SpringOrders.toOrders(table, sort))
                .select(table.fetch(fetcher))
                .execute();
    }
}

File: example/kotlin/jimmer-sql-graphql-kt/repository/src/main/kotlin/org/babyfish/jimmer/sql/example/kt/repository/AuthorRepository.kt
package org.babyfish.jimmer.sql.example.kt.repository

import org.babyfish.jimmer.spring.repository.KRepository
import org.babyfish.jimmer.spring.repository.orderBy
import org.babyfish.jimmer.sql.example.kt.model.Author
import org.babyfish.jimmer.sql.example.kt.model.firstName
import org.babyfish.jimmer.sql.example.kt.model.lastName
import org.babyfish.jimmer.sql.fetcher.Fetcher
import org.babyfish.jimmer.sql.kt.ast.expression.eq
import org.babyfish.jimmer.sql.kt.ast.expression.or
import org.babyfish.jimmer.sql.kt.ast.query.specification.KSpecification
import org.springframework.data.domain.Sort

interface AuthorRepository : KRepository<Author, Long> {

    fun findByName(name: String?): List<Author> =
        sql
            .createQuery(Author::class) {
                name?.let {
                    where(
                        or(
                            table.firstName eq it,
                            table.lastName eq it
                        )
                    )
                }
                select(table)
            }
            .execute()
}
